Exception when reaching last image, and trying to load the next

Thanks for this great module! It's one of the centerpieces of my magic mirror setup, and I really appreciate it!

I'm running into the following error:

**Error description:**

When the module reaches the last of the images in the directory, and attempts to load the next, it will stay on the last image indefinitely. Looking at the logs, an exception is thrown.

**Expected behavior:**

The module should loop and start from the first image in the directory again.

**Details:**

Here's the exception:

> 
> 0|mm       | [2024-11-26 09:26:31.888] [ERROR] TypeError: Cannot read properties of undefined (reading 'name')
> 0|mm       |     at Class.dirChangeAlgo (/home/pi/MagicMirror/modules/MMM-FTP-image/node_helper.js:71:72)
> 0|mm       |     at module.exports.<anonymous> (/home/pi/MagicMirror/modules/MMM-FTP-image/node_helper.js:49:11)
> 0|mm       |     at module.exports.emit (node:events:519:28)
> 0|mm       |     at Object.reentry [as cb] (/home/pi/MagicMirror/modules/MMM-FTP-image/node_modules/ftp/lib/connection.js:238:14)
> 0|mm       |     at Parser.<anonymous> (/home/pi/MagicMirror/modules/MMM-FTP-image/node_modules/ftp/lib/connection.js:117:20)
> 0|mm       |     at Parser.emit (node:events:519:28)
> 0|mm       |     at Parser._write (/home/pi/MagicMirror/modules/MMM-FTP-image/node_modules/ftp/lib/parser.js:59:10)
> 0|mm       |     at writeOrBuffer (node:internal/streams/writable:564:12)
> 0|mm       |     at _write (node:internal/streams/writable:493:10)
> 0|mm       |     at Writable.write (node:internal/streams/writable:502:10)
> 0|mm       |     at Socket.ondata (/home/pi/MagicMirror/modules/MMM-FTP-image/node_modules/ftp/lib/connection.js:273:20)
> 0|mm       |     at Socket.emit (node:events:519:28)
> 0|mm       |     at addChunk (node:internal/streams/readable:559:12)
> 0|mm       |     at readableAddChunkPushByteMode (node:internal/streams/readable:510:3)
> 0|mm       |     at Readable.push (node:internal/streams/readable:390:5)
> 0|mm       |     at TCP.onStreamRead (node:internal/stream_base_commons:191:23)


And here's my config:


>     {
>         module: 'MMM-FTP-image',
>         position: 'middle_center',
>         config: {
>             // FTP server configuration
>             port: 21,
>             user: '<user>',
>             host: '<server>,
>             password: '<password>',
> 
>             // Display configuration
>             opacity: 0.4,
>             height: '100%',
>             imgChangeInterval: 10000,
> 
>             // FTP directory configuration
>             defaultDirPath: '</path/to/images>',
>             dirPathsAuthorized: ['</path/to/images>'],
>         },
>     },


Thank you for your time!

--- a/MMM-FTP-image.js
+++ b/MMM-FTP-image.js
@@ -51,13 +51,24 @@ Module.register('MMM-FTP-image', {
 					this.scheduleImgUpdateInterval();
 					this.finishAllImgInCurrentDirectory = false;
 				}
-				break;
+
+				if (this.imageLoadFinished) break;
 
 			case 'FTP_IMG_BASE64':
 				this.logMessage('Images received !');
 				this.imgBase64 = payload;
 				this.incrementImageIndex();
 				this.updateDom();
+				break;
+
+			case 'RESET':
+				this.logMessage('RESET');
+				this.imageLoadFinished = false;
+				this.finishAllImgInCurrentDirectory = true;
+				this.imgNameList = [];
+				this.imageDisplayedNumber = 0;
+				clearInterval(this.intervalInstance);
+				this.getListImgNameFromFTPServer();
 				break;
 		}
 	},
@@ -149,6 +160,7 @@ Module.register('MMM-FTP-image', {
 
 	incrementImageIndex: function () {
 		this.logMessage(`Current image index: ${this.imageDisplayedNumber}`);
+		console.log('IMG list length', this.imgNameList.length, this.imageDisplayedNumber);
 
 		if (this.imageDisplayedNumber === this.imgNameList.length - 1) {
 			clearInterval(this.intervalInstance);

--- a/MMM-FTP-image.js
+++ b/MMM-FTP-image.js
@@ -160,7 +160,6 @@ Module.register('MMM-FTP-image', {
 
 	incrementImageIndex: function () {
 		this.logMessage(`Current image index: ${this.imageDisplayedNumber}`);
-		console.log('IMG list length', this.imgNameList.length, this.imageDisplayedNumber);
 
 		if (this.imageDisplayedNumber === this.imgNameList.length - 1) {
 			clearInterval(this.intervalInstance);

--- a/node_helper.js
+++ b/node_helper.js
@@ -102,13 +102,20 @@ module.exports = NodeHelper.create({
 
 	moveDir: function (ftp, path) {
 		ftp.cwd(path, function (err) {
-			if (err) throw err;
+			if (err) {
+				console.warn('Error while moving to directory', err);
+				self.reset();
+				throw err;
+			}
 		});
 	},
 
 	sendListName: function (ftp, self) {
 		ftp.list(async function (err, list) {
-			if (err) throw err;
+			if (err) {
+				console.warn('Error while listing files', err);
+				throw err;
+			}
 
 			for (let i = 0; i < list.length; i++) {
 				const file = list[i];
@@ -149,7 +156,12 @@ module.exports = NodeHelper.create({
 	sendBase64Img: async function (ftp, self, payload) {
 		await new Promise((resolve, reject) => {
 			ftp.get(payload.fileName, function (err, stream) {
-				if (err) reject(err);
+				if (err) {
+					console.warn('Error while getting file', err);
+					reject(err);
+
+					self.reset();
+				}
 
 				self.streamToBase64(stream, ftp)
 					.then(function (res) {
@@ -160,6 +172,8 @@ module.exports = NodeHelper.create({
 						resolve();
 					})
 					.catch(function (err) {
+						console.warn('Error while converting stream to base64', err);
+						self.reset();
 						throw new Error(err);
 					});
 			});
@@ -183,16 +197,28 @@ module.exports = NodeHelper.create({
 					ftp.end();
 				})
 				.on('error', error => {
+					console.warn('Error while piping stream', error);
 					reject(error);
+					self.reset();
 				});
 		});
 	},
 
 	getMimeType: function (filename) {
-		for (var s in MimeTypesAuthorized) {
+		for (const s in MimeTypesAuthorized) {
 			if (filename.indexOf(s) === 0) {
 				return MimeTypesAuthorized[s];
 			}
 		}
+	},
+
+	reset: function () {
+		this.dirIndex = 0;
+		this.dirPathVisited = [];
+		this.dirNameList = [];
+		this.imgNameList = [];
+		this.imgBase64 = new Object();
+
+		this.sendSocketNotification('RESET');
 	},
 });

